Add PET volume import feature: GUI, core functions, and realignment

DiellorBasha opened the pull request:
Added support for importing, pre-processing, and registering PET data in Brainstorm, including handling dynamic (multi-frame) PET volumes.

### GUI Update:
- Added an import option for PET in the subject menu.
- Grouped import menu items (MRI, CT, PET) under "Import Volumes" to reduce menu clutter.

### Core Functions:
- New mri_realign function to realign PET frames using SPM, with optional Gaussian smoothing.
- Modified import_mri to support importing and co-registration of dynamic PET volumes.
    - Automatic co-registration and reslicing of PET using SPM after the realignment step.

### User Prompts:
- Allows users to choose whether to align frames and apply smoothing during import.

### Future Work:
- Add support for bulk importing BIDS PET datasets.
- Integrate Freesurfer support for realigning and co-registering PET.

### Requirements and Assumptions:
- PET importing assumes that the corresponding MRI volume has already been imported into the Brainstorm protocol.
- SPM must be installed and configured within Brainstorm to enable PET frame realignment.

### Testing:
- Successfully tested locally using a Brainstorm protocol with .nii PET files.
- Test data available on request.

rcassani commented:
> Test data available on request.

@DiellorBasha, can you send some date for testing?

DiellorBasha commented:
See [link](https://mcgill-my.sharepoint.com/:f:/g/personal/diellor_basha_mcgill_ca/EkI9gewooVdNufSmQ_eIr4UBwx9hJTMvi9UikTuZSbXtNw?email=raymundo.cassani%40mcgill.ca&e=hhk8Wh), containing a Brainstorm protocol and sample PET volumes for import.

rcassani commented on toolbox/anatomy/mri_realign.m:
Please, use indentation for clearer reading

Be careful with the fields (e.g. History and Comment) that are set in the output `sMri` structures but they are not carried on in `import_mri`. Take as example how these are handled in `mrio_coregister` or `mri_reslice`

rcassani commented on toolbox/io/import_mri.m:
```diff
@@ -264,8 +271,25 @@
+                    ' Do you want to align the frames and compute their mean? ' 10 10], ['Dynamic volume']);
+                isSmooth = java_dialog('confirm', ['Do you want to smooth the frames before importing?' 10 10], ['Dynamic volume']);
+                if isRealign && isSmooth
+                    [~, sMri, fileTag] = mri_realign(sMri); % Align frames then register to frame mean
+                elseif isRealign && ~isSmooth
+                    [~, sMri, fileTag] = mri_realign(sMri, [], 0); % Align frames but don't smooth
```
Note that mri_coregister will wipeout the value in the `History` field.
So the History entry added by `mri_realign` is not saved in the final file.

DiellorBasha replied:
Fixed in commit b02066b

rcassani commented on toolbox/anatomy/mri_realign.m:
```diff
@@ -0,0 +1,184 @@
+    if numFrames==1 % If numFrames is 1, volume is static 
+        return
+    else
+        % Remove NaN
+        if any(isnan(sMri.Cube(:)))
+            sMri.Cube(isnan(sMri.Cube)) = 0;
+        end
```
The IF seems unnecessary, why not only the line below?
```
sMri.Cube(isnan(sMri.Cube)) = 0;
```

DiellorBasha replied:
Thanks - removed unneccessary "if" condition in commit 61373f2

rcassani commented on toolbox/anatomy/mri_realign.m:
```diff
@@ -0,0 +1,184 @@
+        % TO DO
+end
+
+% ===== UPDATE HISTORY ========       
```
* This History is not keept in `import_mri
* Same happens for the `Comment`
* Same happens for the `FileTag`. This one should be used inside `import_mri` when creating the name of the volume to save.

rcassani commented on toolbox/anatomy/mri_realign.m:
```diff
@@ -0,0 +1,184 @@
+        error('Invalid call.');
+    end
+% Initialize returned variables
+    sMriAlign = sMri;
```
Are we 100% we want to keep everything? 
That does not sound right, e.g., the `Histogram` will change

DiellorBasha replied:
Now initialized as `sMriAlign = [];` in commit 61373f2.  Histogram and other fields imported during call to `in_mri` in L158.

rcassani reviewed (changes requested):
Overall the code seems to be working well.

There are some minor issues that need to be addressed, see my individual comments. 

@DiellorBasha, I hope you are not very attached to the PET icon, I'll take a lot to it and update it, I find it quite not aesthetic :P

![image](https://github.com/user-attachments/assets/6d31bf2d-b785-46b7-8961-2d43464202ce)

rcassani commented:
@DiellorBasha, PET icon is now updated in `brainstorm.jar` [45b617a](https://github.com/brainstorm-tools/bst-java/commit/45b617a1f9eec939f9a7e65c6f836b4fb0296b73)

![image](https://github.com/user-attachments/assets/e2ad28b5-7faf-4893-b276-bb684d83390a)

sbaillet commented:
❤️ this!

DiellorBasha commented on toolbox/io/import_mri.m:
```diff
@@ -264,8 +271,25 @@
+                elseif isRealign && ~isSmooth
+                    [~, sMri, fileTag] = mri_realign(sMri, [], 0); % Align frames but don't smooth
+                else
+                    sMri.Cube = mean(sMri.Cube, 4);
```
Added optional frame aggregation in a new PET pre-processing dialog in 61373f2

rcassani replied:
In the PET panel, it is a boolean. Why don't we make it a string `'mean'`, so we can easily extend the code to other aggregations (e.g., max) if needed in the future?

DiellorBasha replied:
Agree. Added options for other aggregations through the new `mri_aggregate` function in fad9d5c9c9aa261b2bde3a8a2bace3db7f599bf1. `panel_import_pet` now allows user to choose aggregation method through a combo box: mean, median, max, min, sum, zscore, first frame and last frame 068bbbee7a872fa87cf87402b900ab67283154c4

DiellorBasha replied:
> Why to make mandatory the frame aggregation? Brainstorm can handle 4D Cubes

Note that mri_coregister returns 3D volumes even with 4D inputs.

DiellorBasha commented on toolbox/io/import_mri.m:
```diff
@@ -319,6 +345,10 @@
+            % Reslice PET 
+            elseif isPet
+                isReslice=1; 
+                isInteractive=1;
```
Same as above. In 61373f2, user can now choose whether to co-register and reslice the imported PET volume, with additional choices for registration method ('SPM', 'MNI', 'Ignore').

DiellorBasha commented on toolbox/io/import_mri.m:
Added call to panel_import_pet.m, which consolidates all required choices (alignment, smoothing, aggregation, scaling, registration, reslicing) in one interface. In import_mri,  user options are then routed accordingly to mri_realign and mri_coregister, bypassing the need for 6 separate dialog prompts. See commit 61373f2 for details

rcassani replied:
This is an effective solution.

DiellorBasha commented on toolbox/anatomy/mri_realign.m:
```diff
@@ -0,0 +1,184 @@
+end
+    if isstruct(MriFile) % USAGE: [sMriMean, sMriAlign, fileTag] = mri_realign(sMri, Method)
+        sMri = MriFile; 
+        MriFile = [];
```
Thanks. Removed L65

DiellorBasha commented on toolbox/tree/tree_callbacks.m:
```diff
@@ -1056,8 +1059,9 @@
+                isPet   = strcmpi(nodeType, 'volpet');
                     
-                if (length(bstNodes) == 1)
+                if (length(bstNodes) == 1) && ~isPet
```
Removed ~isPet for now, see commit b02066b.  Also corrected scaling factor for PET data, as it was not being applied on import, giving incorrect uptake values. See https://github.com/brainstorm-tools/brainstorm3/commit/61373f2c5c2a45fdd72b9bf65393f5bc8ac30beb#diff-0f58c25199257166b30995be3635525263396f9a5b136697df058bb35a152c66R581

PET will require its own custom popup in the next PR with options for masking, SUVR, PVC and PET icon without MRI-specific options such as segmentation. Will also update PET colormap in next PR.

DiellorBasha commented:
@rcassani  Thanks for the detailed review and the new PET icon (I was not very attached to the old one). 
 
I've added changes that allow users to choose various processing options during PET import. 
Given the many decisions needed - file selection, alignment, smoothing, aggregation, scaling, registration and reslicing - I've added `panel_import_pet` to handle user choices in a single interface. 
PET can now be imported with/without aligning, aggregating, registering or reslicing. Users can also set the size of the smoothing kernel in mm in the same prompt.  See b02066b, 61373f2, and aaeed20 for this and other minor changes.

rcassani commented on toolbox/io/import_mri.m:
```diff
@@ -264,65 +271,98 @@
+                end
+                % Compute mean across frames if requested
+                if petopts.average
+                    sMri.Cube = mean(sMri.Cube, 4);
```
This average happens after realign, and after smooth right?

The average inside `mri_realign` when does it happen? 
Do we need to have both?

DiellorBasha replied:
For importing, we don't need both - `mri_realign` had option for automatic importing of mean volume generated by SPM Realign for other uses. Removed for this PR.

rcassani reviewed (commented):
@DiellorBasha, there were important problems with the new additions.

1. Problems at importing volumes other than PET
  Solved in: c7d3ac4f
  
2. The repetition of asking user. The idea with the PET panel is good, though it had redundant fields, and did not help skipping steps to the user. Registration was still asked.
  Solved in: fe2f3ba9, f0399b9b, d12a6fda
  
3. Other issues included the non-order of filetag and comment, as this should follow the same order of operations, so, it's another way to see the history of the file. 

A major limitation of the current development is that it cannot be easily scripted, thing that is must. See my detail comments in the code on how overcome this limitation.

rcassani commented on toolbox/io/import_mri.m:
```diff
@@ -285,6 +309,8 @@
             % Ask user to make a choice
             RegMethod = java_dialog('question', [strOptions '<BR><BR></HTML>'], ['Import ', volType], [], cellOptions, 'Reg+reslice');
+        elseif isPet
+            RegMethod = 'SPM'
```
With changes in f0399b9b, the user can directly select the co-registration method in the PET panel

DiellorBasha replied:
Thanks - https://github.com/brainstorm-tools/brainstorm3/commit/f0399b9ba3c8b27f6f78631b70c720bc63eb698f does away with redudant prompts.
I've retained the "registration" check box in  068bbbee7a872fa87cf87402b900ab67283154c4 for more intuitive method selection.

rcassani commented:
@DiellorBasha, also the scaling that happens in the NIfTI files needs to be always done for PET.

https://nipy.org/nibabel/nifti_images.html#data-scaling

DiellorBasha commented:
> @DiellorBasha, also the scaling that happens in the NIfTI files needs to be always done for PET.
> 
> https://nipy.org/nibabel/nifti_images.html#data-scaling

`in_mri_nii` was suppressing scaling by default, even when isScale argument was explicitly set to 1. Fixed in 7d8a88099578ca4f50957a423946c927779811b3. Now applied automatically when importing PET from `mri_realign` through new "isRescale" input argument in `in_mri`, implemented in 700cc92b6814c586b8e517b6d9ae3ff038a8a82c

DiellorBasha commented on toolbox/io/import_mri.m:
```diff
@@ -264,65 +271,98 @@
+                % Registration method
+                RegMethod = petopts.register;
+                % Reslice
+                isReslice = petopts.reslice;
```
I agree and will implement pet_preprocess and masking, SUVR, PVC and other PET-specific processes in coming PRs.  The goal of this PR is to enable importing of PET volumes into Brainstorm in a way that ensures they are correctly aligned and usable for later surface-based analysis.

rcassani replied:
The limitation in the current state is that importing the PET volumes is only available from the GUI as the user needs to set the parameters in the new PET panel.

A way to script importing PET volumes is needed, thus my comment above.

Everything else looks good and ready.

@DiellorBasha, is this non-interactive importing of PET volumes something you can handle?
Otherwise, I can add it myself, though it will not be immediately

DiellorBasha replied:
@rcassani, I folded mri_aggregate and added non-interactive functionalities within mri_realign itself since the PET preprocessing is essentially a call to mri_realign followed by mri_aggregate. I will finalize testing and push the full commit later today.

DiellorBasha replied:
> The limitation in the current state is that importing the PET volumes is only available from the GUI as the user needs to set the parameters in the new PET panel.
> 
> A way to script importing PET volumes is needed, thus my comment above.
> 
> Everything else looks good and ready.
> 
> @DiellorBasha, is this non-interactive importing of PET volumes something you can handle? Otherwise, I can add it myself, though it will not be immediately

@rcassani , it is now possible to pre-process PET non-interactively using `mri_realign`.  A dynamic PET volume can now be realigned with options to smooth and aggregate within `mri_realign`, taking either an Brainstorm MRI structure as input or path to MriFile. This foregoes the need to create a dedicated PET pre-processing script as the main steps are realignment and aggregation.  See commits 24df1c2fa58092880418d9d1d2b2b3bc5c89fceb and the updated `import_mri` in 35edee1bf76fe1557f1a4e0072a4193937f37454.

DiellorBasha commented:
@rcassani - thanks for the improvements. 
I've added options for volume aggregation, rescaling and consolidation of PET pre-processing options as suggested.

The current implementation meets the requirements for importing PET data into Brainstorm and implements the following:
- PET data is realigned, resliced, and coregistered with the subject's default MRI volume
- The imported data is ready for downstream volume and surface-based analysis
- The PET panel simplifies the number of prompts and consolidates key preprocessing decisions 
- Rescaling is automatically applied for imported PET
- Smoothing can be optionally applied during import
- Dynamic PET volumes can be aggregated by mean, median, sum, max, min, zscore 

Proceed as you see fit. I will not be making any further changes to this PR.

rcassani commented on toolbox/tree/tree_callbacks.m:
```diff
@@ -572,10 +572,13 @@
+                    jMenuImportVol= gui_component('Menu', jPopup, [], 'Import volumes', IconLoader.ICON_ANATOMY, [], []);  
+                        gui_component('MenuItem', jMenuImportVol, [], 'MRI', IconLoader.ICON_ANATOMY, [], @(h,ev)bst_call(@import_mri, iSubject, [], [], 1));
+                        gui_component('MenuItem', jMenuImportVol, [], 'CT', IconLoader.ICON_VOLCT, [], @(h,ev)bst_call(@import_mri, iSubject, [], [], 1, 1, 'Import CT'));
+                        gui_component('MenuItem', jMenuImportVol, [], 'PET', IconLoader.ICON_VOLPET, [], @(h,ev)bst_call(@import_mri, iSubject, [], [], 0, 1, 'Import PET'));
```
Fully reverted in dfe67fadb1ae5ab07fc5f9c7454ddae8633c9909

rcassani commented:
@DiellorBasha, thanks for the updates

rcassani reviewed (approved):
@DiellorBasha, this is ready to be merged!

- Added the popup menu for PET, and its options.
  This will a place for further PET tools
